### PR TITLE
Convert to enzyme for shallow rendering

### LINKED ISSUE #2508: Remove core/ components that are unused
Now that #2373 is done there are some components hanging out in `core/` like `NavBar` that can be removed as they're entirely unused. We should go through and delete them. Faster tests, less cruft! 🎉

### LINKED ISSUE #2539: Use Enzyme for shallow rendering
We have our own shallow rendering helper, but with enzyme we can move to enzyme's `shallow()`.

### PR DESCRIPTION
**Pull request by tofumatt**
Got interested in this while shallow-testing a new component and decided to see how much effort it would take to replace all our `shallowRender()` calls with Enzyme's `shallow()`. Turns out it wasn't too much so here we go.

Annoyingly the `yarn.lock` additions don't highlight this for the like 200 removed lines this actually is. 😆

This relates to #1104 but I'm not sure if we should consider it closed as that issue also mentions replacing things like `Simulate.click()` which this patch doesn't do. Maybe we should close that issue as it says "Consider Enzyme" which we clearly did by merging this, but make new issues for specific other tweaks we want?

Fixes #2508.
Fixes #2539.

### CODE REVIEW

File: tests/unit/amo/components/TestSearchPage.js [hunk 1]
```diff
@@ -48,44 +48,49 @@ describe('<SearchPage />', () => {
 
   it('renders a Paginate', () => {
     const root = render();
-    const paginator = findByTag(root, Paginate);
-    expect(paginator.props.count).toEqual(80);
-    expect(paginator.props.currentPage).toEqual(3);
-    expect(paginator.props.pathname).toEqual('/search/');
-    expect(paginator.props.queryParams).toEqual({ q: 'foo' });
+    const paginator = root.find(Paginate);
+    expect(paginator.prop('count')).toEqual(80);
```
Comment (muffinresearch) [inline tests/unit/amo/components/TestSearchPage.js]: Simple value checks should probably use `toBe()` even though `toEqual()` will work fine. `toBe()` is just using `===` whereas `toEqual()` is more complicated to allow for checking object properties etc..

Comment (muffinresearch) [inline tests/unit/amo/components/TestSearchPage.js]: I like this concise summary of the differences from https://stackoverflow.com/a/25130805/156158

>    expect(x).toEqual(y); compares objects or primitives x and y and passes if they are equivalent
    expect(x).toBe(y); compares objects or primitives x and y and passes if they are the same object

Comment (tofumatt) [inline tests/unit/amo/components/TestSearchPage.js]: Oh, I didn't know that. I didn't like that the two things *seem* like they're the same thing and didn't like using both in the same test. I guess `toBe` is faster?

File: tests/unit/core/components/TestInstallButton.js [hunk 2]
```diff
@@ -37,7 +38,7 @@ describe('<InstallButton />', () => {
   });
 
   const render = (props) =>
-    shallowRender(<InstallButtonBase {...renderProps(props)} />);
+    shallow(<InstallButtonBase {...renderProps(props)} />);
 
   const renderToDom = (props) => findDOMNode(
```
Comment (muffinresearch) [inline tests/unit/core/components/TestInstallButton.js]: Presumably this could use mount? http://airbnb.io/enzyme/docs/api/mount.html

Comment (tofumatt) [inline tests/unit/core/components/TestInstallButton.js]: oh yeah probably but I tried to avoid moving anything that wasn't `shallowRender`. I didn't want the diff to be too large and figured we can do another PR for `findDOMNode`.

File: tests/unit/core/components/TestInstallButton.js [hunk 3]
```diff
@@ -80,13 +87,13 @@ describe('<InstallButton />', () => {
     const root = renderToDom({ addon, installTheme, status: UNKNOWN });
 
     const preventDefault = sinon.spy();
-    const buttonComponent = root.querySelector('.InstallButton-button');
-    Simulate.click(buttonComponent, { preventDefault });
+    const button = root.querySelector('.InstallButton-button');
+    Simulate.click(button, { preventDefault });
```
Comment (muffinresearch) [inline tests/unit/core/components/TestInstallButton.js]: There's a simulate method in enzyme for both `mount()` and `shallow()`

File: tests/unit/core/components/TestInstallButton.js [hunk 4]
```diff
@@ -201,7 +216,9 @@ describe('<InstallButton />', () => {
 
     Simulate.click(installButton);
 
-    expect(fakeLog.info.firstCall.args[0]).toEqual('Adding OpenSearch Provider');
-    expect(fakeWindow.external.AddSearchProvider.firstCall.args[0]).toEqual(fakeAddon.installURL);
+    expect(fakeLog.info.firstCall.args[0]).toEqual(
```
Comment (muffinresearch) [inline tests/unit/core/components/TestInstallButton.js]: If you want to split expect calls onto multiple lines putting the `.toEqual` on the next line might be more natural.

File: tests/unit/amo/components/TestSearchPage.js (hunk 1 above)
Comment (muffinresearch) [inline tests/unit/amo/components/TestSearchPage.js]: Not sure re: perf but it's possible.

Comment (muffinresearch) [inline tests/unit/amo/components/TestSearchPage.js]: I agree re: them sounding similar. It's quite hard to stop just using `toEqual()` everywhere.

File: tests/unit/amo/components/TestFeaturedAddons.js [hunk 5]
```diff
@@ -20,25 +21,23 @@ describe('<FeaturedAddons />', () => {
...
+    expect(root.childAt(0).contains('More Featured Extensions')).toBeTruthy();
   });
 
   it('renders a FeaturedAddons page with themes HTML', () => {
     store.dispatch(
       featuredActions.getFeatured({ addonType: ADDON_TYPE_THEME }));
     const root = render(mapStateToProps(store.getState()));
 
-    expect(root.props.children[0].props.children).toContain('More Featured Themes');
+    expect(root.childAt(0).contains('More Featured Themes')).toBeTruthy();
```
Comment (muffinresearch) [inline tests/unit/amo/components/TestFeaturedAddons.js]: Does this give as good error messages if it faile compared to toContain?

File: tests/unit/core/components/TestInstallButton.js [hunk 6]
```diff
@@ -101,17 +108,18 @@ describe('<InstallButton />', () => {
...
-    const buttonComponent = root.props.children[1];
-    expect(buttonComponent.type).toEqual(Button);
-
-    const props = buttonComponent.props;
-    expect(props.children).toEqual('Add to Firefox');
-    expect(props.className).toContain('InstallButton-button');
-    expect(props.className).not.toContain('Button--small');
-    expect(props.href).toEqual(installURL);
+    expect(root.type()).toEqual('div');
+    expect(root.hasClass('InstallButton')).toBeTruthy();
```
Comment (muffinresearch) [inline tests/unit/core/components/TestInstallButton.js]: This should probably use https://github.com/blainekasten/enzyme-matchers#tohaveclassnameclassnamestring since you have jest-enzyme.

**Comment by muffinresearch** [comment]
I missed this first time around - there's a few places that could use the jest-enzyme matchers. Mostly this will be better for exposing decent error messaging when an assertion fails. See this for the list https://github.com/blainekasten/enzyme-matchers#assertions

**Comment by kumar303** [comment]
> This relates to #1104 but I'm not sure if we should consider it closed as that issue also mentions replacing things like Simulate.click() which this patch doesn't do. Maybe

I think you can close it and then file a new issue for replacing the calls to `Simulate.click()`

**Comment by tofumatt** [comment]
> I missed this first time around - there's a few places that could use the jest-enzyme matchers. Mostly this will be better for exposing decent error messaging when an assertion fails. See this for the list https://github.com/blainekasten/enzyme-matchers#assertions

Oh nice, I'll do that. I did as close to a 1:1 conversion here but I'll try those.

File: tests/unit/core/components/TestInstallButton.js (hunk 3 above)
Comment (muffinresearch) [inline tests/unit/core/components/TestInstallButton.js]: Should have read the description - this is fine as a follow up.

File: tests/unit/amo/components/TestFeaturedAddons.js [hunk 7]
```diff
@@ -20,25 +21,23 @@ describe('<FeaturedAddons />', () => {
...
+    return shallow(<FeaturedAddonsBase i18n={getFakeI18nInst()} {...props} />);
   }
 
   it('renders a FeaturedAddons page with no add-ons set', () => {
     store.dispatch(
       featuredActions.getFeatured({ addonType: ADDON_TYPE_EXTENSION }));
     const root = render(mapStateToProps(store.getState()));
 
-    expect(root.props.children[0].props.children).toContain('More Featured Extensions');
+    expect(root.childAt(0).contains('More Featured Extensions')).toBeTruthy();
```
Comment (kumar303) [inline tests/unit/amo/components/TestFeaturedAddons.js]: This should be [`expect(root.childAt(0)).toIncludeText('More Featured Extensions')`](https://github.com/blainekasten/enzyme-matchers#toincludetexttextstring).

File: tests/unit/amo/components/TestFeaturedAddons.js (hunk 5 above)
Comment (kumar303) [inline tests/unit/amo/components/TestFeaturedAddons.js]: It should be [`expect(root.childAt(0)).toIncludeText('More Featured Themes')`](https://github.com/blainekasten/enzyme-matchers#toincludetexttextstring).

File: tests/unit/core/components/TestInstallButton.js [hunk 8]
```diff
@@ -50,28 +51,34 @@ describe('<InstallButton />', () => {
...
-    expect(buttonComponent.props['data-browsertheme']).toEqual(JSON.stringify(themePreview.getThemeData(addon)));
+
+    expect(root.type()).toEqual('div');
+    expect(root.hasClass('InstallButton')).toBeTruthy();
+    expect(root.hasClass('InstallButton--use-button')).toBeTruthy();
+
+    const button = root.childAt(1);
+
+    expect(button.type()).toEqual(Button);
+    expect(button.contains('Install Theme')).toBeTruthy();
```
Comment (kumar303) [inline tests/unit/core/components/TestInstallButton.js]: This should be [`expect(button).toIncludeText('Install Theme')`](https://github.com/blainekasten/enzyme-matchers#toincludetexttextstring).

File: tests/unit/core/components/TestInstallButton.js [hunk 9]
```diff
@@ -50,28 +51,34 @@ describe('<InstallButton />', () => {
...
     const root = render({ hasAddonManager: false, addon });
-    expect(root.type).toEqual('div');
-    expect(root.props.className).toEqual('InstallButton InstallButton--use-button');
-    const buttonComponent = root.props.children[1];
-    expect(buttonComponent.type).toEqual(Button);
-    expect(buttonComponent.props.children).toEqual('Install Theme');
-    expect(buttonComponent.props['data-browsertheme']).toEqual(JSON.stringify(themePreview.getThemeData(addon)));
+
+    expect(root.type()).toEqual('div');
+    expect(root.hasClass('InstallButton')).toBeTruthy();
```
Comment (kumar303) [inline tests/unit/core/components/TestInstallButton.js]: This should use [`expect(...).toHaveClassName`](https://github.com/blainekasten/enzyme-matchers#tohaveclassnameclassnamestring) so we get a better failure message.

File: tests/unit/core/components/TestInstallButton.js [hunk 10]
```diff
@@ -50,28 +51,34 @@ describe('<InstallButton />', () => {
...
+    expect(root.hasClass('InstallButton--use-button')).toBeTruthy();
```
Comment (kumar303) [inline tests/unit/core/components/TestInstallButton.js]: Same thing here about [`expect(...).toHaveClassName`](https://github.com/blainekasten/enzyme-matchers#tohaveclassnameclassnamestring).

File: tests/unit/core/components/TestInstallButton.js [hunk 11]
```diff
@@ -101,17 +108,18 @@ describe('<InstallButton />', () => {
...
-    expect(props.href).toEqual(installURL);
+    expect(root.type()).toEqual('div');
+    expect(root.hasClass('InstallButton')).toBeTruthy();
+    expect(root.hasClass('InstallButton--use-button')).toBeTruthy();
+
+    const button = root.childAt(1);
+
+    expect(button.type()).toEqual(Button);
+
+    expect(button.contains('Add to Firefox')).toBeTruthy();
```
Comment (kumar303) [inline tests/unit/core/components/TestInstallButton.js]: This should be [`expect(button).toIncludeText('Add to Firefox')`](https://github.com/blainekasten/enzyme-matchers#toincludetexttextstring).

File: tests/unit/core/components/TestInstallButton.js [hunk 12]
```diff
@@ -101,17 +108,18 @@ describe('<InstallButton />', () => {
...
+    expect(button.hasClass('InstallButton-button')).toBeTruthy();
```
Comment (kumar303) [inline tests/unit/core/components/TestInstallButton.js]: Same thing about `toHaveClassName`

File: tests/unit/core/components/TestInstallButton.js [hunk 13]
```diff
@@ -101,17 +108,18 @@ describe('<InstallButton />', () => {
...
+    expect(button.hasClass('Button--small')).toBeFalsy();
```
Comment (kumar303) [inline tests/unit/core/components/TestInstallButton.js]: also `expect(...).not.toHaveClassName()`

File: tests/unit/core/components/TestInstallButton.js [hunk 14]
```diff
@@ -125,37 +133,40 @@ describe('<InstallButton />', () => {
...
-    expect(buttonComponent.type).toEqual(Button);
-    expect(buttonComponent.props.className).toContain(
-      'InstallButton-button--disabled');
-    expect(buttonComponent.props.href).toBe(installURL);
+    expect(root.type()).toEqual('div');
+
+    const button = root.childAt(1);
+
+    expect(button.type()).toEqual(Button);
+    expect(button.hasClass('InstallButton-button--disabled')).toBeTruthy();
```
Comment (kumar303) [inline tests/unit/core/components/TestInstallButton.js]: `expect(...).toHaveClassName()`

File: tests/unit/core/components/TestInstallButton.js [hunk 15]
```diff
@@ -167,12 +178,14 @@ describe('<InstallButton />', () => {
...
-    expect(buttonComponent.type).toEqual(Button);
-    expect(buttonComponent.props.className).toContain('Button--action');
-    expect(buttonComponent.props.className).toContain('InstallButton-button');
-    expect(buttonComponent.props.children).toEqual('Add to Firefox');
+    expect(root.type()).toEqual('div');
+
+    const button = root.childAt(1);
+
+    expect(button.type()).toEqual(Button);
+    expect(button.hasClass('Button--action')).toBeTruthy();
```
Comment (kumar303) [inline tests/unit/core/components/TestInstallButton.js]: `expect(...).toHaveClassName()`

File: tests/unit/core/components/TestInstallButton.js [hunk 16]
```diff
@@ -167,12 +178,14 @@ describe('<InstallButton />', () => {
...
+    expect(button.hasClass('InstallButton-button')).toBeTruthy();
```
Comment (kumar303) [inline tests/unit/core/components/TestInstallButton.js]: `expect(...).toHaveClassName()`

File: tests/unit/core/components/TestInstallButton.js [hunk 17]
```diff
@@ -167,12 +178,14 @@ describe('<InstallButton />', () => {
...
+    expect(button.contains('Add to Firefox')).toBeTruthy();
```
Comment (kumar303) [inline tests/unit/core/components/TestInstallButton.js]: `expect(button).toIncludeText(...)`

File: tests/unit/core/components/TestInstallButton.js [hunk 18]
```diff
@@ -181,11 +194,13 @@ describe('<InstallButton />', () => {
...
-    const buttonComponent = root.props.children[1];
-    expect(buttonComponent.type).toEqual(Button);
-    expect(buttonComponent.props.className).toContain('InstallButton-button--disabled');
-    expect(buttonComponent.props.children).toEqual('Add to Firefox');
+    expect(root.type()).toEqual('div');
+
+    const button = root.childAt(1);
+
+    expect(button.type()).toEqual(Button);
+    expect(button.hasClass('InstallButton-button--disabled')).toBeTruthy();
```
Comment (kumar303) [inline tests/unit/core/components/TestInstallButton.js]: `toHaveClassName()`

File: tests/unit/core/components/TestInstallButton.js [hunk 19]
```diff
@@ -181,11 +194,13 @@ describe('<InstallButton />', () => {
...
+    expect(button.contains('Add to Firefox')).toBeTruthy();
```
Comment (kumar303) [inline tests/unit/core/components/TestInstallButton.js]: `.toIncludeText()`

File: tests/unit/ui/components/TestButton.js [hunk 20]
```diff
@@ -1,31 +1,32 @@
...
+    const button = shallow(
       <Button className="Foo" onClick={onClick}>My button!</Button>);
 
-    expect(button.type).toEqual('button');
-    expect(button.props.className).toContain('Button');
-    expect(button.props.className).toContain('Foo');
-    expect(button.props.onClick).toBe(onClick);
-    expect(button.props.children).toEqual('My button!');
+    expect(button.type()).toEqual('button');
+    expect(button.hasClass('Button')).toBeTruthy();
```
Comment (kumar303) [inline tests/unit/ui/components/TestButton.js]: `toHaveClassName`

File: tests/unit/ui/components/TestButton.js [hunk 21]
```diff
@@ -1,31 +1,32 @@
...
+    expect(button.hasClass('Foo')).toBeTruthy();
```
Comment (kumar303) [inline tests/unit/ui/components/TestButton.js]: `toHaveClassName`

File: tests/unit/ui/components/TestButton.js [hunk 22]
```diff
@@ -1,31 +1,32 @@
...
+    expect(button.prop('onClick')).toEqual(onClick);
+    expect(button.contains('My button!')).toBeTruthy();
```
Comment (kumar303) [inline tests/unit/ui/components/TestButton.js]: `toIncludeText()`

File: tests/unit/ui/components/TestButton.js [hunk 23]
```diff
@@ -1,31 +1,32 @@
...
+    const button = shallow(
       <Button className="Bar" to={href}>Link text!</Button>);
 
-    expect(button.type).toEqual(Link);
-    expect(button.props.className).toContain('Button');
-    expect(button.props.className).toContain('Bar');
-    expect(button.props.to).toBe(href);
-    expect(button.props.children).toEqual('Link text!');
+    expect(button.type()).toEqual(Link);
+    expect(button.hasClass('Button')).toBeTruthy();
```
Comment (kumar303) [inline tests/unit/ui/components/TestButton.js]: `toHaveClassName`

File: tests/unit/ui/components/TestButton.js [hunk 24]
```diff
@@ -1,31 +1,32 @@
...
+    expect(button.hasClass('Bar')).toBeTruthy();
```
Comment (kumar303) [inline tests/unit/ui/components/TestButton.js]: `toHaveClassName`

File: tests/unit/ui/components/TestButton.js [hunk 25]
```diff
@@ -1,31 +1,32 @@
...
+    expect(button.prop('to')).toBe(href);
+    expect(button.contains('Link text!')).toBeTruthy();
```
Comment (kumar303) [inline tests/unit/ui/components/TestButton.js]: `toIncludeText`

File: tests/unit/ui/components/TestIcon.js [hunk 26]
```diff
@@ -1,16 +1,22 @@
...
-import { shallowRender } from 'tests/unit/helpers';
+
 
 describe('<Icon />', () => {
   it('maps the name to a className', () => {
-    const root = shallowRender(<Icon name="foo" />);
-    expect(root.props.className).toEqual('Icon Icon-foo');
+    const root = shallow(<Icon name="foo" />);
+
+    expect(root.hasClass('Icon Icon-foo')).toBeTruthy();
```
Comment (kumar303) [inline tests/unit/ui/components/TestIcon.js]: `toHaveClassName`

File: tests/unit/ui/components/TestIcon.js [hunk 27]
```diff
@@ -1,16 +1,22 @@
...
+    expect(root.hasClass('Icon Icon-foo')).toBeTruthy();
```
Comment (kumar303) [inline tests/unit/ui/components/TestIcon.js]: `toHaveClassName`

File: tests/unit/ui/components/TestIcon.js [hunk 28]
```diff
@@ -1,16 +1,22 @@
...
   });
 
   it('allows a custom className', () => {
-    const root = shallowRender(<Icon name="bar" className="sup" />);
-    expect(root.props.className).toEqual('Icon Icon-bar sup');
+    const root = shallow(<Icon name="bar" className="sup" />);
+
+    expect(root.hasClass('Icon')).toBeTruthy();
```
Comment (kumar303) [inline tests/unit/ui/components/TestIcon.js]: `toHaveClassName`

File: tests/unit/ui/components/TestIcon.js [hunk 29]
```diff
@@ -1,16 +1,22 @@
...
+    expect(root.hasClass('Icon-bar')).toBeTruthy();
```
Comment (kumar303) [inline tests/unit/ui/components/TestIcon.js]: `toHaveClassName`

File: tests/unit/ui/components/TestIcon.js [hunk 30]
```diff
@@ -1,16 +1,22 @@
...
+    expect(root.hasClass('sup')).toBeTruthy();
```
Comment (kumar303) [inline tests/unit/ui/components/TestIcon.js]: `toHaveClassName`

File: tests/unit/amo/components/TestSearchPage.js (hunk 1 above)
Comment (tofumatt) [inline tests/unit/amo/components/TestSearchPage.js]: `toEqual` seems to still to strict checks for things like `true` and `80` though, right? Is there a reason not to use it for checks against simple types like this? Will `expect(80).toEqual('80');` pass? If not it'd be nice to always use the same API for equality checks. I always hated having to switch between `.equal()`, `.deepEqual()`, and `.strictEqual()` in the past...

File: tests/unit/core/components/TestInstallButton.js (hunk 4 above)
Comment (tofumatt) [inline tests/unit/core/components/TestInstallButton.js]: Works for me.

File: src/core/components/NavBar/index.js [hunk 31]
```diff
@@ -1,38 +0,0 @@
-import React from 'react';
```
Comment (tofumatt) [inline src/core/components/NavBar/index.js]: There's an issue open for it this PR marks as fixing... I caught it awhile back. I'm just doing it in this PR because that component's tests used our old `shallowRender` so I deleted it here rather than fixing its tests only to delete it later 😄

File: tests/setup.js [hunk 32]
```diff
@@ -3,6 +3,9 @@ import config from 'config';
 import areIntlLocalesSupported from 'intl-locales-supported';
 import values from 'object.values';
 
+import 'jest-enzyme';
```
Comment (tofumatt) [inline tests/setup.js]: Agreed. Will do.

Comment (tofumatt) [inline tests/setup.js]: Actually, just checking now it seems that was only required for earlier versions of Jest. I removed that line and all the tests still work/pass. I haven't used much/any of the `jest-enzyme` integration yet though, if it's needed for that I'll keep it in but add a comment to explain.

Comment (tofumatt) [inline tests/setup.js]: Aha, just checked add stuff like `.toIncludeText('foo')` needs this side-effect import to monkey-patch all the `expect()` objects.

I've added some docs around it to explain why we do it.

File: tests/unit/amo/components/TestSearchPage.js (hunk 1 above)
Comment (muffinresearch) [inline tests/unit/amo/components/TestSearchPage.js]: > Will expect(80).toEqual('80'); pass?

No you'd get: 

repl_it_-_jest_compiler

`toBe` checks identity through using `===` so if you were comparing two objects it would only pass if the objects were identical.

The second thing is the complexity in the code behind `isEqual()` see https://github.com/facebook/jest/blob/master/packages/jest-matchers/src/jasmine-utils.js#L58 which is the function behind it. Whether this amounts to anything that would actually result in a meaningful slowdown remains to be seen.

From the perspective of just running up the tests using `toEqual()` most of the time wouldn't cause problems. Maybe we don't worry about it?

Comment (muffinresearch) [inline tests/unit/amo/components/TestSearchPage.js]: > I always hated having to switch between .equal(), .deepEqual(), and .strictEqual() in the past...

In general expect via jest means that `.equal()` and `deepEqual()` are replaced by `toEqual()` and `.strictEqual()` is replaced by `toBe()` so we could work on the basis of only using `toBe()` when we need the strict equality + identity check - leaving any potential perf issues aside from now - and then we can look at that separately if it matters.
